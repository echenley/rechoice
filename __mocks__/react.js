const React = require.requireActual('react')

React.Fragment = 'Fragment'

module.exports = React
