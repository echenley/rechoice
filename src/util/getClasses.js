// @flow

import type { ClassMap } from '../types'

export default (classes: $Shape<ClassMap>) => ({
  disabled: 'disabled',
  focused: 'focused',
  input: 'rechoice-input',
  inputMirror: 'rechoice-input-mirror',
  inputWrapper: 'rechoice-input-wrapper',
  item: 'rechoice-item',
  menu: 'rechoice-menu',
  root: 'rechoice',
  selected: 'selected',
  tags: 'rechoice-tags',
  value: 'rechoice-value',
  ...classes,
})
