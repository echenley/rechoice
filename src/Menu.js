// @flow

import * as React from 'react'
import cx from 'classnames'
import DefaultMenu from './DefaultMenu'
import type {
  ClassMap,
  Item as ItemType,
  ItemRenderer,
  MenuRenderer,
  LoadingState,
} from './types'

type Props = {
  classes: ClassMap,
  controlsId: string,
  focusedItemIndex: ?number,
  getItemId: (i: number) => string,
  inputValue: string,
  isItemSelected: (item: ItemType) => boolean,
  isOpen: boolean,
  itemKey: string,
  items: Array<ItemType>,
  loadingIndicator?: React.Node,
  onHoverItem: (i: number) => mixed,
  onSelect: (item: ItemType) => mixed,
  renderItem: ItemRenderer,
  renderMenu?: MenuRenderer,
  focusedItemRef: (ref: ?HTMLElement) => mixed,
  menuRef: (ref: ?HTMLElement) => mixed,
  state?: LoadingState,
}

class Menu extends React.Component<Props> {
  renderItems = () => {
    const {
      classes,
      focusedItemIndex,
      getItemId,
      inputValue,
      isItemSelected,
      itemKey,
      items,
      onHoverItem,
      onSelect,
      renderItem,
      focusedItemRef,
    } = this.props

    return items.map((item, i) => {
      const isSelected = isItemSelected(item)
      const isFocused =
        typeof focusedItemIndex === 'number' && i === focusedItemIndex

      const renderedItem = renderItem({
        inputValue,
        isFocused,
        isSelected,
        item,
      })

      const itemClasses = cx(classes.item, renderedItem.props.className, {
        [classes.focused]: isFocused,
        [classes.selected]: isSelected,
      })

      return React.cloneElement(renderedItem, {
        'aria-selected': isFocused,
        className: itemClasses,
        id: getItemId(i),
        key: item[itemKey],
        onClick: e => {
          if (renderedItem.props.onClick) {
            renderedItem.props.onClick(e)
          }

          onSelect(item)
        },
        onMouseEnter: () => onHoverItem(i),
        ref: isFocused ? focusedItemRef : null,
        role: 'option',
      })
    })
  }

  render() {
    const {
      classes,
      controlsId,
      inputValue,
      isOpen,
      loadingIndicator,
      renderMenu = DefaultMenu,
      menuRef,
      state = 'success',
    } = this.props

    const menu = renderMenu({
      children: this.renderItems(),
      className: classes.menu,
      listProps: {
        id: controlsId,
        role: 'listbox',
      },
      inputValue,
      isOpen,
      loadingIndicator,
      menuRef,
      state,
    })

    if (!menu) {
      return menu
    }

    if (menu.type === 'string') {
      return React.cloneElement(React.Children.only(menu), {
        ref: menuRef,
      })
    }

    return menu
  }
}

export default Menu
