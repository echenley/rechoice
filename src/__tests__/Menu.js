import * as React from 'react'
import { shallow } from 'enzyme'
import getClasses from '../util/getClasses'
import Menu from '../Menu'

const item3 = { id: 3 }

// prettier-ignore
const items = [
  { id: 1 },
  { id: 2 },
  item3,
  { id: 4 },
  { id: 5 },
]

const renderComponent = props => {
  return shallow(
    <Menu
      classes={getClasses()}
      controlsId="controls-id"
      focusedItemIndex={2}
      getItemId={i => `item-${i}`}
      inputValue="test"
      isItemSelected={item => item.id === 4}
      isOpen={true}
      itemKey="id"
      items={items}
      loadingIndicator={<div className="loading-indicator" />}
      onHoverItem={() => {}}
      onSelect={() => {}}
      renderItem={({ item }) => <div>{item.id}</div>}
      setFocusedItemRef={() => {}}
      setMenuRef={() => {}}
      state="success"
      {...props}
    />
  )
}

describe('Menu', () => {
  it('renders correctly', () => {
    expect(renderComponent().getElement()).toMatchSnapshot()
  })

  it('calls onSelect when an item is clicked', () => {
    const onSelect = jest.fn()
    const component = renderComponent({ onSelect })
    component.find('#item-0').simulate('click')
    expect(onSelect).toHaveBeenCalledWith({ id: 1 })
  })

  it('calls onHoverItem when item is moused over', () => {
    const onHoverItem = jest.fn()
    const component = renderComponent({ onHoverItem })
    component.find('#item-2').simulate('mouseenter')
    expect(onHoverItem).toHaveBeenCalledWith(2)
  })
})
