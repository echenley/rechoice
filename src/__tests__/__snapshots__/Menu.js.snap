// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu renders correctly 1`] = `
<div
  className="rechoice-menu"
>
  <ul
    id="controls-id"
    role="listbox"
  >
    <div
      aria-selected={false}
      className="rechoice-item"
      id="item-0"
      onClick={[Function]}
      onMouseEnter={[Function]}
      role="option"
    >
      1
    </div>
    <div
      aria-selected={false}
      className="rechoice-item"
      id="item-1"
      onClick={[Function]}
      onMouseEnter={[Function]}
      role="option"
    >
      2
    </div>
    <div
      aria-selected={true}
      className="rechoice-item focused"
      id="item-2"
      onClick={[Function]}
      onMouseEnter={[Function]}
      role="option"
    >
      3
    </div>
    <div
      aria-selected={false}
      className="rechoice-item selected"
      id="item-3"
      onClick={[Function]}
      onMouseEnter={[Function]}
      role="option"
    >
      4
    </div>
    <div
      aria-selected={false}
      className="rechoice-item"
      id="item-4"
      onClick={[Function]}
      onMouseEnter={[Function]}
      role="option"
    >
      5
    </div>
  </ul>
</div>
`;
